compiler/rustc_middle/src/hir/map/mod.rs

```diff
@@ -159,12 +158,12 @@ impl<'hir> Map<'hir> {
         }
     }
 
-    pub fn items(self) -> impl Iterator<Item = &'hir Item<'hir>> + 'hir {
-        let krate = self.krate();
-        krate.owners.iter().filter_map(|owner| match owner.as_owner()?.node() {
-            OwnerNode::Item(item) => Some(item),
-            _ => None,
-        })
+    pub fn items(self) -> impl Iterator<Item = ItemId> + 'hir {
+        self.tcx.hir_crate_items(()).items.iter().map(|id| *id)
+    }
+
+    pub fn par_items(self, f: impl Fn(ItemId) + Sync + Send) {
+        par_for_each_in(self.tcx.hir_crate_items(()).items.to_vec(), f);
```

Comment:
`to_vec` allocates a new vector needlessly.

```suggestion
        par_for_each_in(&self.tcx.hir_crate_items(()).items[..], |id| f(*id));
```

Reply:
Could you do something similar to `par_for_each_module` and only keep one version?

Reply:
I'm going to rename this method.

And do you mean this?
```rust
    pub fn for_each_module(self, f: impl Fn(LocalDefId)) {
        let crate_items = self.tcx.hir_crate_items(());
        for module in crate_items.submodules.iter() {
            f(*module)
        }
    }

    #[cfg(not(parallel_compiler))]
    #[inline]
    pub fn par_for_each_module(self, f: impl Fn(LocalDefId)) {
        self.for_each_module(f)
    }
```
If so, we would need a `for_each_item` method. We might end up with something like this?

```rust
    pub fn for_each_item(self, f: impl Fn(ItemId)) {
....
    }

    #[cfg(not(parallel_compiler))]
    #[inline]
    pub fn par_for_each_item(self, f: impl Fn(ItemId)) {
        self.for_each_item(f)
    }

    #[cfg(parallel_compiler)]
    #[inline]
    pub fn par_for_each_item(self, f: impl Fn(ItemId)) {
        par_for_each_in(...)
    }
```

but `par_for_each_in` has conditional compilation depending on `parallel_compiler` which would go unused.
Sorry, If I'm misunderstanding something. And maybe we can work on this parallelization on a follow up PR?